azext_iot/iothub/providers/device_identity.py

```diff
@@ -0,0 +1,437 @@
+        while len(all_hub_devices) < len(config.devices) and scope_retries < MAX_DEVICE_SCOPE_RETRIES:
+            sleep(3)
+            scope_retries += 1
+            logger.info("Retrying device scope query - attempt {} of {}"
+                        .format(scope_retries, MAX_DEVICE_SCOPE_RETRIES))
+            all_hub_devices = _execute_query(query_args, query_method)
+
+        if len(all_hub_devices) < len(config.devices):
+            raise AzureResponseError(
+                "An error occurred - Failed to fetch device scopes for all devices after {} retries"
+                .format(scope_retries)
+            )
+
+        # set all device scopes
+        scope_dict: Dict[str, str] = {}
+        for device in all_hub_devices:
+            id = device["deviceId"]
+            if device_config_dict.get(id, None):
+                scope_dict[id] = device["deviceScope"]
+
+        # Set parent / child relationships
+        device_to_parent_iterator = (
+            tqdm(device_to_parent_dict, desc="Setting device parents")
+            if visualize
+            else device_to_parent_dict
+        )
+        for device_id in device_to_parent_iterator:
+            # get device properties
+            device = self.service_sdk.devices.get_identity(id=device_id)
+            parent_id = device_to_parent_dict[device_id]
+            parent_scope = scope_dict[parent_id]
+            # set new parent scope
+            device.parent_scopes = [parent_scope]
+            # update device
+            self.service_sdk.devices.create_or_update_identity(
+                id=device_id, device=device, if_match="*"
+            )
+
+        # update edge config / set-modules
+        devices_config_iterator = (
+            tqdm(config.devices, desc="Setting edge module content")
+            if visualize
+            else config.devices
+        )
+        for device_config in devices_config_iterator:
+            device_id = device_config.device_id
+            deployment_content = device_config.deployment
+            if deployment_content:
+                self.service_sdk.configuration.apply_on_edge_device(
+                    id=device_id, content=deployment_content
+                )
+
+        # Print device bundle details after other visuals
+        if bundle_output_directory:
+            num_bundles = len(config.devices)
+            bundle_plural = '' if num_bundles == 1 else 's'
+            print(f"{num_bundles} device bundle{bundle_plural} created in folder: {abspath(bundle_output_directory)}")
+
+    def delete_device_identities(self, device_ids: List[str], confirm: bool = False):
+        for id in device_ids:
+            try:
+                self.service_sdk.devices.delete_identity(id=id, if_match="*")
+            except Exception as err:
+                raise AzureResponseError(err)
+        if confirm:
+            existing_devices = self.service_sdk.devices.get_devices()
+            if len(existing_devices):
```

Comment:
minor thing - `list` with an item is True (and empty list is False) so you could do `if existing_devices:`